### PR TITLE
LSP: differentiate diagnostic underline by severity

### PR DESCRIPTION
**Pull request by pwntester**

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -660,8 +660,19 @@ do
       local start = diagnostic.range.start
       local finish = diagnostic.range["end"]
 
+      local hl = underline_highlight_name
+      if diagnostic.severity == protocol.DiagnosticSeverity.Error then
+          hl = underline_highlight_name..'Error'
```
Comment (bfredl) [inline runtime/lua/vim/lsp/util.lua]: You will also need to add defaults like at 550. Probably use default link:

    hi default link LspDiagnosticsUnderlineError LspDiagnosticsUnderline

Comment (pwntester) [reply runtime/lua/vim/lsp/util.lua]: thanks! fixed in new commit

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -548,6 +548,9 @@ do
 
   local underline_highlight_name = "LspDiagnosticsUnderline"
   api.nvim_command(string.format("highlight default %s gui=underline cterm=underline", underline_highlight_name))
+  for kind, _ in pairs(protocol.DiagnosticSeverity) do
+    api.nvim_command(string.format("highlight default link %s%s %s", underline_highlight_name, kind, underline_highlight_name))
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua]: can use `vim.cmd` for these now.

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -660,8 +663,19 @@ do
...
 
+      local hl = underline_highlight_name
+      if diagnostic.severity == protocol.DiagnosticSeverity.Error then
+          hl = underline_highlight_name..'Error'
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Warning then
+          hl = underline_highlight_name..'Warning'
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Information then
+          hl = underline_highlight_name..'Information'
+      elseif diagnostic.severity == protocol.DiagnosticSeverity.Hint then
+          hl = underline_highlight_name..'Hint'
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua]: 2-space indent.

File: runtime/lua/vim/lsp/util.lua [hunk 4]
```diff
@@ -660,8 +663,19 @@ do
...
+      end
+
       -- TODO care about encoding here since this is in byte index?
-      highlight_range(bufnr, diagnostic_ns, underline_highlight_name,
+      highlight_range(bufnr, diagnostic_ns, hl,
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua]: the if-else verbosity can be avoided by using a map (which may later be useful as part of the `protocol` module if it doesn't already exist...


```
local hlmap = {
  [protocol.DiagnosticSeverity.Error]='Error',
  [protocol.DiagnosticSeverity.Warning]='Warning',
  [protocol.DiagnosticSeverity.Information]='Information',
  [protocol.DiagnosticSeverity.Hint]='Hint',
}
highlight_range(bufnr, diagnostic_ns,
  underline_highlight_name..hlmap[diagnostic.severity],
    {start.line, start.character},
    {finish.line, finish.character}
)
```

**Comment by justinmk** [comment]
I pushed an update with docs and a basic test that checks for the existence of the highlight groups:

```
  LspDiagnosticsError
  LspDiagnosticsHint
  LspDiagnosticsInformation
  LspDiagnosticsUnderline
  LspDiagnosticsUnderline1
  LspDiagnosticsUnderline2
  LspDiagnosticsUnderline3
  LspDiagnosticsUnderline4
  LspDiagnosticsUnderlineError
  LspDiagnosticsUnderlineHint
  LspDiagnosticsUnderlineInformation
  LspDiagnosticsUnderlineWarning
  LspDiagnosticsWarning
```

This seems like an absurd number of extra highlight groups and I assume we're going to add even more. I don't know what all of these are for but it seems excessive to me. What user is going to spend time fiddling with dozens of new highlight groups?

Anyways, I'll merge this after CI finishes.

**Comment by pwntester** [comment]
LspDiagnosticsUnderline1
LspDiagnosticsUnderline2
LspDiagnosticsUnderline3
LspDiagnosticsUnderline4

should not be defined. The problem is that table is reversed in protocol:297 and iterating pairs(protocol.DiagnosticSeverity) returns both the table items and the reversed version. Is there anyway to prevent this? otherwise, we need to duplicate DiagnosticsSeverity in util.lua

**Comment by justinmk** [comment]
> table is reversed in protocol:297 and iterating pairs(protocol.DiagnosticSeverity) returns both the table items and the reversed version. Is there anyway to prevent this?

@pwntester good note. I added a check for `type(..) == 'string'`. That's good enough for now.

**Pull request merged**